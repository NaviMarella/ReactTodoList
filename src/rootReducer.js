export const ActionTypes = {
    TOGGLE_SIDEBAR: "TOGGLE_SIDEBAR",
    SET_CURRENT_LABEL: "SET_CURRENT_LABEL",
    ADD_TODO: "ADD_TODO",
    EDIT_TODO: "EDIT_TODO",
    DELETE_TODO: "DELETE_TODO",
    ADD_LABEL: "ADD_LABEL",
    DELETE_LABEL: "DELETE_LABEL"
};

export const RootReducer = (state = initialState, action) => {
    switch (action.type) {
        case "TOGGLE_SIDEBAR":
            return {
                ...state,
                hideSidebar: !state.hideSidebar
            }
        case "SET_CURRENT_LABEL":
            return {
                ...state,
                currentLabel: action.data
            }
        case "ADD_TODO":
            return {
                ...state,
                todoList: [
                    ...state.todoList,
                    {
                        id: getUniq(),
                        label: state.currentLabel,
                        text: action.data
                    }
                ]
            }
        case "EDIT_TODO":
            return {
                ...state,
                todoList: state.todoList.map(
                    content => (content.id === action.payload.id) ? { ...content,
                        text: action.payload.text
                    } :
                    content
                )
            }
        case "DELETE_TODO":
            return {
                ...state,
                todoList: state.todoList.filter(content => content.id !== action.payload.id)
            }
        case "ADD_LABEL":
            return {
                ...state,
                currentLabel: action.data,
                labels: [
                    ...state.labels,
                    {
                        id: getUniq(),
                        name: action.data
                    }
                ]
            }
        case "DELETE_LABEL":
        return {
            ...state,
            currentLabel: state.currentLabel === action.data ? "TODOS" : state.currentLabel,
            labels: state.labels.filter(label => label.name !== action.data),
            todoList: state.todoList.filter(todo => todo.label !== action.data)
        }
        case "TOGGLE_COMPLETE":
            return {
                ...state,
                todoList: state.todoList.map(
                    content => (content.id === action.payload.id) ? { ...content,
                        isCompleted: !content.isCompleted
                    } :
                    content
                )
            }
        default:
            return state;
    }
};

export const getUniq = function () {
    return Math.random().toString(36).substr(2, 9);
}

export const initialState = {
    hideSidebar: false,
    currentLabel: "TODOS",
    labels: [{
            id: 1,
            name: 'TODOS'
        },
        {
            id: getUniq(),
            name: 'Work'
        },
        {
            id: getUniq(),
            name: 'Groceries'
        }
    ],
    todoList: [{
            id: getUniq(),
            label: 'TODOS',
            text: "Learn React",
            isCompleted: true
        }, {
            id: getUniq(),
            label: 'Groceries',
            text: "Buy Vegetables",
            isCompleted: true
        }, {
            id: getUniq(),
            label: 'Work',
            text: "Learn Redux",
            isCompleted: false
        },
        {
            id: 4,
            label: 'Work',
            text: "Learn Jest",
            isCompleted: false
        }
    ]
};
