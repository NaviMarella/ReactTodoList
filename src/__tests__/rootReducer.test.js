import React from 'react';
import { RootReducer, initialState } from "../rootReducer";


describe('Root Reducer Tests', () => {
  test('render default state snapshot', () => {
    const action = { type: "" };
    const store = RootReducer(initialState, action);
    expect(store).toMatchSnapshot();
  });

  test('render custom state snapshot', () => {
      const action = { type: "" };
      const store = RootReducer({id: 1, name: test}, action);
      expect(store).toMatchSnapshot();
  });

  test('toggle sidebar action snapshot', () => {
      const action = { type: "TOGGLE_SIDEBAR" };
      const store = RootReducer(initialState, action);
      expect(store).toMatchSnapshot();
  });

  test('set current label action snapshot', () => {
      const action = { type: "SET_CURRENT_LABEL", data: "" };
      const store = RootReducer(initialState, action);
      expect(store).toMatchSnapshot();
  });

   test('edit todo action snapshot', () => {
      const action = { type: "EDIT_TODO", payload: {id: 1, text: "test"} };
      const store = RootReducer(initialState, action);
      expect(store).toMatchSnapshot();
  });

   test('add todo action snapshot', () => {
      const action = { type: "ADD_TODO", data: "test" };
      const store = RootReducer(initialState, action);
      expect(store).toMatchSnapshot();
  });

  test('delete todo action snapshot', () => {
      const action = { type: "DELETE_TODO", payload: {id: 1} };
      const store = RootReducer(initialState, action);
      expect(store).toMatchSnapshot();
  });

  test('add label action snapshot', () => {
      const action = { type: "ADD_LABEL", data: "test" };
      const store = RootReducer(initialState, action);
      expect(store).toMatchSnapshot();
  });

  test('delete label action snapshot', () => {
      const action = { type: "DELETE_LABEL", data: "test" };
      const store = RootReducer(initialState, action);
      expect(store).toMatchSnapshot();
  });

  test('toogle complete action snapshot', () => {
      const action = { type: "TOGGLE_COMPLETE", payload: {id: 1} };
      const store = RootReducer(initialState, action);
      expect(store).toMatchSnapshot();
  });
});
