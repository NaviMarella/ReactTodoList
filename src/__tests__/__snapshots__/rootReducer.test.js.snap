// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Root Reducer Tests add label action snapshot 1`] = `
Object {
  "currentLabel": "test",
  "hideSidebar": false,
  "labels": Array [
    Object {
      "id": 1,
      "name": "TODOS",
    },
    Object {
      "id": "hvkrsbnk1",
      "name": "Work",
    },
    Object {
      "id": "sm9z95wy8",
      "name": "Groceries",
    },
    Object {
      "id": "20w48qgyi",
      "name": "test",
    },
  ],
  "todoList": Array [
    Object {
      "id": "vrpkrk5ek",
      "isCompleted": true,
      "label": "TODOS",
      "text": "Learn React",
    },
    Object {
      "id": "512k2gpo0",
      "isCompleted": true,
      "label": "Groceries",
      "text": "Buy Vegetables",
    },
    Object {
      "id": "l9xit5mv6",
      "isCompleted": false,
      "label": "Work",
      "text": "Learn Redux",
    },
    Object {
      "id": 4,
      "isCompleted": false,
      "label": "Work",
      "text": "Learn Jest",
    },
  ],
}
`;

exports[`Root Reducer Tests add todo action snapshot 1`] = `
Object {
  "currentLabel": "TODOS",
  "hideSidebar": false,
  "labels": Array [
    Object {
      "id": 1,
      "name": "TODOS",
    },
    Object {
      "id": "hvkrsbnk1",
      "name": "Work",
    },
    Object {
      "id": "sm9z95wy8",
      "name": "Groceries",
    },
  ],
  "todoList": Array [
    Object {
      "id": "vrpkrk5ek",
      "isCompleted": true,
      "label": "TODOS",
      "text": "Learn React",
    },
    Object {
      "id": "512k2gpo0",
      "isCompleted": true,
      "label": "Groceries",
      "text": "Buy Vegetables",
    },
    Object {
      "id": "l9xit5mv6",
      "isCompleted": false,
      "label": "Work",
      "text": "Learn Redux",
    },
    Object {
      "id": 4,
      "isCompleted": false,
      "label": "Work",
      "text": "Learn Jest",
    },
    Object {
      "id": "3lr53hk1c",
      "label": "TODOS",
      "text": "test",
    },
  ],
}
`;

exports[`Root Reducer Tests delete label action snapshot 1`] = `
Object {
  "currentLabel": "TODOS",
  "hideSidebar": false,
  "labels": Array [
    Object {
      "id": 1,
      "name": "TODOS",
    },
    Object {
      "id": "hvkrsbnk1",
      "name": "Work",
    },
    Object {
      "id": "sm9z95wy8",
      "name": "Groceries",
    },
  ],
  "todoList": Array [
    Object {
      "id": "vrpkrk5ek",
      "isCompleted": true,
      "label": "TODOS",
      "text": "Learn React",
    },
    Object {
      "id": "512k2gpo0",
      "isCompleted": true,
      "label": "Groceries",
      "text": "Buy Vegetables",
    },
    Object {
      "id": "l9xit5mv6",
      "isCompleted": false,
      "label": "Work",
      "text": "Learn Redux",
    },
    Object {
      "id": 4,
      "isCompleted": false,
      "label": "Work",
      "text": "Learn Jest",
    },
  ],
}
`;

exports[`Root Reducer Tests delete todo action snapshot 1`] = `
Object {
  "currentLabel": "TODOS",
  "hideSidebar": false,
  "labels": Array [
    Object {
      "id": 1,
      "name": "TODOS",
    },
    Object {
      "id": "hvkrsbnk1",
      "name": "Work",
    },
    Object {
      "id": "sm9z95wy8",
      "name": "Groceries",
    },
  ],
  "todoList": Array [
    Object {
      "id": "vrpkrk5ek",
      "isCompleted": true,
      "label": "TODOS",
      "text": "Learn React",
    },
    Object {
      "id": "512k2gpo0",
      "isCompleted": true,
      "label": "Groceries",
      "text": "Buy Vegetables",
    },
    Object {
      "id": "l9xit5mv6",
      "isCompleted": false,
      "label": "Work",
      "text": "Learn Redux",
    },
    Object {
      "id": 4,
      "isCompleted": false,
      "label": "Work",
      "text": "Learn Jest",
    },
  ],
}
`;

exports[`Root Reducer Tests edit todo action snapshot 1`] = `
Object {
  "currentLabel": "TODOS",
  "hideSidebar": false,
  "labels": Array [
    Object {
      "id": 1,
      "name": "TODOS",
    },
    Object {
      "id": "hvkrsbnk1",
      "name": "Work",
    },
    Object {
      "id": "sm9z95wy8",
      "name": "Groceries",
    },
  ],
  "todoList": Array [
    Object {
      "id": "vrpkrk5ek",
      "isCompleted": true,
      "label": "TODOS",
      "text": "Learn React",
    },
    Object {
      "id": "512k2gpo0",
      "isCompleted": true,
      "label": "Groceries",
      "text": "Buy Vegetables",
    },
    Object {
      "id": "l9xit5mv6",
      "isCompleted": false,
      "label": "Work",
      "text": "Learn Redux",
    },
    Object {
      "id": 4,
      "isCompleted": false,
      "label": "Work",
      "text": "Learn Jest",
    },
  ],
}
`;

exports[`Root Reducer Tests render custom state snapshot 1`] = `
Object {
  "id": 1,
  "name": [Function],
}
`;

exports[`Root Reducer Tests render default state snapshot 1`] = `
Object {
  "currentLabel": "TODOS",
  "hideSidebar": false,
  "labels": Array [
    Object {
      "id": 1,
      "name": "TODOS",
    },
    Object {
      "id": "hvkrsbnk1",
      "name": "Work",
    },
    Object {
      "id": "sm9z95wy8",
      "name": "Groceries",
    },
  ],
  "todoList": Array [
    Object {
      "id": "vrpkrk5ek",
      "isCompleted": true,
      "label": "TODOS",
      "text": "Learn React",
    },
    Object {
      "id": "512k2gpo0",
      "isCompleted": true,
      "label": "Groceries",
      "text": "Buy Vegetables",
    },
    Object {
      "id": "l9xit5mv6",
      "isCompleted": false,
      "label": "Work",
      "text": "Learn Redux",
    },
    Object {
      "id": 4,
      "isCompleted": false,
      "label": "Work",
      "text": "Learn Jest",
    },
  ],
}
`;

exports[`Root Reducer Tests set current label action snapshot 1`] = `
Object {
  "currentLabel": "",
  "hideSidebar": false,
  "labels": Array [
    Object {
      "id": 1,
      "name": "TODOS",
    },
    Object {
      "id": "hvkrsbnk1",
      "name": "Work",
    },
    Object {
      "id": "sm9z95wy8",
      "name": "Groceries",
    },
  ],
  "todoList": Array [
    Object {
      "id": "vrpkrk5ek",
      "isCompleted": true,
      "label": "TODOS",
      "text": "Learn React",
    },
    Object {
      "id": "512k2gpo0",
      "isCompleted": true,
      "label": "Groceries",
      "text": "Buy Vegetables",
    },
    Object {
      "id": "l9xit5mv6",
      "isCompleted": false,
      "label": "Work",
      "text": "Learn Redux",
    },
    Object {
      "id": 4,
      "isCompleted": false,
      "label": "Work",
      "text": "Learn Jest",
    },
  ],
}
`;

exports[`Root Reducer Tests toggle sidebar action snapshot 1`] = `
Object {
  "currentLabel": "TODOS",
  "hideSidebar": true,
  "labels": Array [
    Object {
      "id": 1,
      "name": "TODOS",
    },
    Object {
      "id": "hvkrsbnk1",
      "name": "Work",
    },
    Object {
      "id": "sm9z95wy8",
      "name": "Groceries",
    },
  ],
  "todoList": Array [
    Object {
      "id": "vrpkrk5ek",
      "isCompleted": true,
      "label": "TODOS",
      "text": "Learn React",
    },
    Object {
      "id": "512k2gpo0",
      "isCompleted": true,
      "label": "Groceries",
      "text": "Buy Vegetables",
    },
    Object {
      "id": "l9xit5mv6",
      "isCompleted": false,
      "label": "Work",
      "text": "Learn Redux",
    },
    Object {
      "id": 4,
      "isCompleted": false,
      "label": "Work",
      "text": "Learn Jest",
    },
  ],
}
`;

exports[`Root Reducer Tests toogle complete action snapshot 1`] = `
Object {
  "currentLabel": "TODOS",
  "hideSidebar": false,
  "labels": Array [
    Object {
      "id": 1,
      "name": "TODOS",
    },
    Object {
      "id": "hvkrsbnk1",
      "name": "Work",
    },
    Object {
      "id": "sm9z95wy8",
      "name": "Groceries",
    },
  ],
  "todoList": Array [
    Object {
      "id": "vrpkrk5ek",
      "isCompleted": true,
      "label": "TODOS",
      "text": "Learn React",
    },
    Object {
      "id": "512k2gpo0",
      "isCompleted": true,
      "label": "Groceries",
      "text": "Buy Vegetables",
    },
    Object {
      "id": "l9xit5mv6",
      "isCompleted": false,
      "label": "Work",
      "text": "Learn Redux",
    },
    Object {
      "id": 4,
      "isCompleted": false,
      "label": "Work",
      "text": "Learn Jest",
    },
  ],
}
`;
